client/service/src/chain_ops.rs

```diff
@@ -116,55 +303,36 @@ impl<
 			let client = &self.client;
 			let queue = &mut self.import_queue;
 
-			// Start by reading the number of blocks if not done so already.
-			let count = match count {
-				Some(c) => c,
-				None => {
-					let c: u64 = match Decode::decode(&mut io_reader_input) {
-						Ok(c) => c,
-						Err(err) => {
-							let err = format!("Error reading file: {}", err);
-							return std::task::Poll::Ready(Err(From::from(err)));
-						},
-					};
-					info!("📦 Importing {} blocks", c);
-					count = Some(c);
-					c
-				}
-			};
-
 			// Read blocks from the input.
-			if read_block_count < count {
-				match SignedBlock::<Self::Block>::decode(&mut io_reader_input) {
-					Ok(signed) => {
-						let (header, extrinsics) = signed.block.deconstruct();
-						let hash = header.hash();
-						// import queue handles verification and importing it into the client
-						queue.import_blocks(BlockOrigin::File, vec![
-							IncomingBlock::<Self::Block> {
-								hash,
-								header: Some(header),
-								body: Some(extrinsics),
-								justification: signed.justification,
-								origin: None,
-								allow_missing_state: false,
-								import_existing: force,
-							}
-						]);
+			match block_iter.next() {
+				None => {
+					let read_block_count = block_iter.read_block_count();
+					let num_expected_blocks = block_iter.num_expected_blocks();
+
+					if importing_is_done(num_expected_blocks, read_block_count, link.imported_blocks) {
```

Comment:
This was in the code previously. `importing_is_done` checks that the number of blocks processed by the link is the same as the expected number of blocks. Not sure how this works if we return early, and don't let the time for the link to finish importing blocks. I believe it will be dropped, right?

Reply:
This is related to [one of my previous comments](https://github.com/paritytech/substrate/pull/5816#discussion_r416793180) too :D

Reply:
Your `Iterator` implementation only returns `None`, after  all blocks are read? So, we don't need this check here.

Reply:
Removed in https://github.com/paritytech/substrate/pull/5816/commits/377823c0a648a3eb2e61185a257a61023067893d :)